tests/unit/sanakoyeu_et_al_2018/test_loss.py

```diff
@@ -0,0 +1,152 @@
+
+        with subtests.test(mode=mode):
+            test_op.real() if mode else test_op.fake()
+            actual = test_op(input_image)
+            prediction = encoder(input_image) * 2.0
+            desired = torch.mean(prediction)
+            ptu.assert_allclose(actual, desired)
+            desired_accuracy = torch.mean(comparator(prediction, 0.0).float())
+            ptu.assert_allclose(test_op.accuracy, desired_accuracy)
+
+
+def test_PredictionOperator_call(subtests, input_image):
+    encoder = enc.SequentialEncoder((nn.Conv2d(3, 3, 1),))
+    predictor = nn.Conv2d(3, 3, 1)
+
+    op = paper.PredictionOperator(encoder, predictor)
+
+    for mode in (True, False):
+        comparator = torch.ge if mode else torch.lt
+
+        with subtests.test(mode=mode):
+            op.real() if mode else op.fake()
+            actual = op(input_image)
+            prediction = predictor(encoder(input_image))
+            with subtests.test("loss"):
+                desired = torch.mean(
+                    binary_cross_entropy_with_logits(
+                        prediction,
+                        torch.ones_like(prediction)
+                        if mode
+                        else torch.zeros_like(prediction),
+                    )
+                )
+                ptu.assert_allclose(actual, desired)
+            with subtests.test("accuracy"):
+                desired_accuracy = torch.mean(comparator(prediction, 0.0).float())
+                ptu.assert_allclose(op.accuracy, desired_accuracy)
+
+
+def test_MultiLayerPredictionOperator(subtests, input_image):
+    class TestOperator(paper.EncodingDiscriminatorOperator):
+        def input_enc_to_repr(self, image):
+            return image * 2.0
+
+        def calculate_score(self, input_repr):
+            self.accuracy = self.calculate_accuracy(input_repr)
+            return torch.mean(input_repr)
+
+        def calculate_accuracy(self, input_repr: torch.Tensor) -> torch.Tensor:
+            comparator = torch.ge if self._target_distribution else torch.lt
+            return torch.mean(comparator(input_repr, 0.0).float())
+
+    def get_encoding_op(encoder, score_weight):
+        return TestOperator(encoder, score_weight)
+
+    layers = [str(index) for index in range(3)]
+    modules = [(layer, nn.Conv2d(3, 3, 1)) for layer in layers]
+    multi_layer_encoder = enc.MultiLayerEncoder(modules)
+
+    multi_layer_prediction_op = paper.MultiLayerPredictionOperator(
+        multi_layer_encoder, layers, get_encoding_op
+    )
+
+    for mode in (True, False):
+        multi_layer_prediction_op.real() if mode else multi_layer_prediction_op.fake()
+        _ = multi_layer_prediction_op(input_image)
+        with subtests.test(mode=mode):
+            with subtests.test("operator mode"):
+                for op in multi_layer_prediction_op.discriminator_operators():
+                    assert op._target_distribution == mode
+
+            with subtests.test("accuracy"):
```

Comment:
You should almost never test for private attributes: since they might change without warning this test might also break.

```suggestion
```